<a href="https://colab.research.google.com/github/w15d0177/AIFFEL_quest_cr/blob/main/MainQuest/Quest02/pokemon_machine_learning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [163]:
# 드라이브 마운트하기
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [164]:
# 라이브러리 가져오기
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline
%config InlineBackend.figure_format = 'retina'

In [165]:
# 데이터셋을 pandas로 불러오기
import os

csv_path = '/content/drive/MyDrive/AIFFEL_2024/data/Pokemon.csv'
original_data = pd.read_csv(csv_path)

# **(1) 기본 모델이 동작하는 코드**

In [166]:
# 원본 데이터 확인하기 - 포켓몬 800마리 및 각 포켓몬의 13가지 특성
print(original_data.shape)
original_data.head()

(800, 13)


#                   Name Type 1  Type 2  Total  HP  Attack  Defense  \
0  1              Bulbasaur  Grass  Poison    318  45      49       49   
1  2                Ivysaur  Grass  Poison    405  60      62       63   
2  3               Venusaur  Grass  Poison    525  80      82       83   
3  3  VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
4  4             Charmander   Fire     NaN    309  39      52       43   

   Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0       65       65     45           1      False  
1       80       80     60           1      False  
2      100      100     80           1      False  
3      122      120     80           1      False  
4       60       50     65           1      False

In [167]:
# 원본 데이터의 컬럼 확인하기 - 13가지 특성
original_data.columns

Index(['#', 'Name', 'Type 1', 'Type 2', 'Total', 'HP', 'Attack', 'Defense',
       'Sp. Atk', 'Sp. Def', 'Speed', 'Generation', 'Legendary'],
      dtype='object')

In [168]:
# 의미 없는 컬럼 #, 문자열을 포함하는 컬럼 (Name, Type1, Type2), 타겟 데이터터인 Legendary 제외하기
# 나머지는 배열로 features 변수에 담기
features = ['Total', 'HP', 'Attack', 'Defense', 'Sp. Atk', 'Sp. Def', 'Speed', 'Generation']

In [169]:
# 타켓 데이터인 Legendary를 target 변수에 담기
target = 'Legendary'

In [170]:
# features의 컬럼들에 해당하는 데이터를 변수 X에 저장하고 확인하기 - 포켓몬 800마리 및 각 포켓몬의 8가지 특성
X = original_data[features]
print(X.shape)
X.head()

(800, 8)


Total  HP  Attack  Defense  Sp. Atk  Sp. Def  Speed  Generation
0    318  45      49       49       65       65     45           1
1    405  60      62       63       80       80     60           1
2    525  80      82       83      100      100     80           1
3    625  80     100      123      122      120     80           1
4    309  39      52       43       60       50     65           1

In [171]:
# target의 컬럼에 해당하는 데이터를 변수 y에 저장하고 확인하기 - 총 800마리 포켓몬 - 전설의 포켓몬 또는 일반 포켓문
y = original_data[target]
print(y.shape)
y.head()

(800,)


0    False
1    False
2    False
3    False
4    False
Name: Legendary, dtype: bool

In [172]:
# 훈련 데이터와 테스트 데이터 분리하기 - 훈련 데이터 640 - 테스트 데이터 160
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=15)

print(X_train.shape, y_train.shape)
print(X_test.shape, y_test.shape)

(640, 8) (640,)
(160, 8) (160,)


In [173]:
# 의사결정 트리 모델을 변수 model에 저장하기
from sklearn.tree import DecisionTreeClassifier

model = DecisionTreeClassifier(random_state=25)
model

DecisionTreeClassifier(random_state=25)

In [174]:
# 훈련 데이터로 모델 훈련시키기
model.fit(X_train, y_train)

# 테스트 데이터를 통해 예측값 얻기
y_pred = model.predict(X_test)

In [175]:
# 예측값 (y_pred) 및 정답 (y_test) 확인하기
from sklearn.metrics import confusion_matrix
confusion_matrix(y_test, y_pred)

array([[144,   3],
       [  5,   8]])

## **✅ 작업 수행 결과**

            Positive: Legendary=True (전설의 포켓몬)    
            Negative: Legendary=False (일반 포켓몬)

                - TN (True Negative): 144   
                    일반 포켓몬을 일반 포켓몬이라고 알맞게 판단     

                - FP (False Positive): 3   
                    일반 포켓몬을 전설의 포켓몬이라고 잘못 판단     

                - FN (False Negative): 5    
                    전설의 포켓몬을 일반 포켓몬이라고 잘못 판단     
                     
                - TP (True Positive): 8   
                    전설의 포켓몬을 전설의 포켓몬이라고 알맞게 판단
   
- *총 160개의 테스트 데이터 중 8개만 잘못 판단하여 95%의 높은 정확도 (accuracy)를 달성했다.*    
- *그러나, 800마리 포켓몬 중 735마리가 일반 포켓몬인 상황에서 모델이 아무런 학습을 안 하고 모든 답을 일반 포켓몬 하나로 답해도 92%의 정확도를 달성할 수 있다.*           
- *따라서, 해당 데이터셋에서는 정확도로 모델의 성능을 평가하기에는 무리가 있다.*

In [176]:
# 다른 척도로 모델의 성능 평가하기
from sklearn.metrics import classification_report
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

       False       0.97      0.98      0.97       147
        True       0.73      0.62      0.67        13

    accuracy                           0.95       160
   macro avg       0.85      0.80      0.82       160
weighted avg       0.95      0.95      0.95       160



## **✅ 작업 수행 결과**

- *해당 데이터는 전설의 포켓몬 65마리이고 일반 포켓몬 735마리로 구성되어 불균형하다.*
- *따라서, 적은 양의 positive 데이터인 전설의 포켓몬을 잡아내는 것이 중요하다.*
- *다시 말하면, FN (False Negative)를 줄여야 한다.*
- *그러나, FN이 분모에 위치한 Recall의 값 (0.62)이 낮은 것으로 보아, FN 값이 높다고 보여진다.*
- *전설의 포켓몬이 확연히 적은 불균형한 데이터에서 전설의 포켓몬을 잡아내지 못하는 오류가 크다는 것은 바람직하지 않다고 보여진다.*
- *이는 앞서 보인 높은 정확도가 불균형한 데이터에서 경계되어야 함을 보여주기도 하다.*

# **(2) 데이터 전처리**

In [178]:
# 원본 데이터를 복사하여 pokemon 변수에 저장하고, pandas로 데이터 확인하기 - 포켓몬 800마리 및 각 포켓몬의 13가지 특성
pokemon = original_data.copy()
print(pokemon.shape)
pokemon.head()

(800, 13)


#                   Name Type 1  Type 2  Total  HP  Attack  Defense  \
0  1              Bulbasaur  Grass  Poison    318  45      49       49   
1  2                Ivysaur  Grass  Poison    405  60      62       63   
2  3               Venusaur  Grass  Poison    525  80      82       83   
3  3  VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
4  4             Charmander   Fire     NaN    309  39      52       43   

   Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0       65       65     45           1      False  
1       80       80     60           1      False  
2      100      100     80           1      False  
3      122      120     80           1      False  
4       60       50     65           1      False

In [181]:
# "전설의 포켓몬" 데이터셋 확인하기 - 전설의 포켓몬 65마리 및 각 포켓몬의 13가지 특성
legendary = pokemon[pokemon["Legendary"] == True].reset_index(drop=True)
print(legendary.shape)
legendary.head()

(65, 13)


#                 Name    Type 1    Type 2  Total   HP  Attack  Defense  \
0  144             Articuno       Ice    Flying    580   90      85      100   
1  145               Zapdos  Electric    Flying    580   90      90       85   
2  146              Moltres      Fire    Flying    580   90     100       90   
3  150               Mewtwo   Psychic       NaN    680  106     110       90   
4  150  MewtwoMega Mewtwo X   Psychic  Fighting    780  106     190      100   

   Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0       95      125     85           1       True  
1      125       90    100           1       True  
2      125       85     90           1       True  
3      154       90    130           1       True  
4      154      100    130           1       True

In [183]:
# 일반 포켓몬 데이터셋 확인하기 - 일반 포켓몬 735마리 및 각 포켓몬의 13가지 특성
ordinary = pokemon[pokemon["Legendary"] == False].reset_index(drop=True)
print(ordinary.shape)
ordinary.head()

(735, 13)


#                   Name Type 1  Type 2  Total  HP  Attack  Defense  \
0  1              Bulbasaur  Grass  Poison    318  45      49       49   
1  2                Ivysaur  Grass  Poison    405  60      62       63   
2  3               Venusaur  Grass  Poison    525  80      82       83   
3  3  VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
4  4             Charmander   Fire     NaN    309  39      52       43   

   Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0       65       65     45           1      False  
1       80       80     60           1      False  
2      100      100     80           1      False  
3      122      120     80           1      False  
4       60       50     65           1      False

In [184]:
# 전체 컬럼 이해하기 - 13개의 컬럼 - 13가지 특성
print(len(pokemon.columns))
pokemon.columns

13


Index(['#', 'Name', 'Type 1', 'Type 2', 'Total', 'HP', 'Attack', 'Defense',
       'Sp. Atk', 'Sp. Def', 'Speed', 'Generation', 'Legendary'],
      dtype='object')

In [185]:
# 컬럼 name_count 추가하기 - 이름의 길이 확인
pokemon["name_count"] = pokemon["Name"].apply(lambda i: len(i))
pokemon.head()

#                   Name Type 1  Type 2  Total  HP  Attack  Defense  \
0  1              Bulbasaur  Grass  Poison    318  45      49       49   
1  2                Ivysaur  Grass  Poison    405  60      62       63   
2  3               Venusaur  Grass  Poison    525  80      82       83   
3  3  VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
4  4             Charmander   Fire     NaN    309  39      52       43   

   Sp. Atk  Sp. Def  Speed  Generation  Legendary  name_count  
0       65       65     45           1      False           9  
1       80       80     60           1      False           7  
2      100      100     80           1      False           8  
3      122      120     80           1      False          21  
4       60       50     65           1      False          10

In [186]:
# 컬럼 long_name 추가하기 - 이름의 길이가 10이상이면 True, 미만이면 False
pokemon["long_name"] = pokemon["name_count"] >= 10
pokemon.head()

#                   Name Type 1  Type 2  Total  HP  Attack  Defense  \
0  1              Bulbasaur  Grass  Poison    318  45      49       49   
1  2                Ivysaur  Grass  Poison    405  60      62       63   
2  3               Venusaur  Grass  Poison    525  80      82       83   
3  3  VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
4  4             Charmander   Fire     NaN    309  39      52       43   

   Sp. Atk  Sp. Def  Speed  Generation  Legendary  name_count  long_name  
0       65       65     45           1      False           9      False  
1       80       80     60           1      False           7      False  
2      100      100     80           1      False           8      False  
3      122      120     80           1      False          21       True  
4       60       50     65           1      False          10       True

In [187]:
# 컬럼 Name_nospace 추가하기 - 띄어쓰기 없이 이름의 길이 확인
pokemon["Name_nospace"] = pokemon["Name"].apply(lambda i: i.replace(" ", ""))
pokemon.tail()

#                 Name   Type 1 Type 2  Total  HP  Attack  Defense  \
795  719              Diancie     Rock  Fairy    600  50     100      150   
796  719  DiancieMega Diancie     Rock  Fairy    700  50     160      110   
797  720  HoopaHoopa Confined  Psychic  Ghost    600  80     110       60   
798  720   HoopaHoopa Unbound  Psychic   Dark    680  80     160       60   
799  721            Volcanion     Fire  Water    600  80     110      120   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  name_count  long_name  \
795      100      150     50           6       True           7      False   
796      160      110    110           6       True          19       True   
797      150      130     70           6       True          19       True   
798      170      130     80           6       True          18       True   
799      130       90     70           6       True           9      False   

           Name_nospace  
795             Diancie  
796  DiancieMegaDiancie  
797  HoopaHoopaConfined  
798   HoopaHoopaUnbound  
799           Volcanion

In [188]:
# 컬럼 name_isalpha 추가하기 - 이름이 알파벳으로만 이루어졌는지 확인
pokemon["name_isalpha"] = pokemon["Name_nospace"].apply(lambda i: i.isalpha())
pokemon.head()

#                   Name Type 1  Type 2  Total  HP  Attack  Defense  \
0  1              Bulbasaur  Grass  Poison    318  45      49       49   
1  2                Ivysaur  Grass  Poison    405  60      62       63   
2  3               Venusaur  Grass  Poison    525  80      82       83   
3  3  VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
4  4             Charmander   Fire     NaN    309  39      52       43   

   Sp. Atk  Sp. Def  Speed  Generation  Legendary  name_count  long_name  \
0       65       65     45           1      False           9      False   
1       80       80     60           1      False           7      False   
2      100      100     80           1      False           8      False   
3      122      120     80           1      False          21       True   
4       60       50     65           1      False          10       True   

           Name_nospace  name_isalpha  
0             Bulbasaur          True  
1               Ivysaur          True  
2              Venusaur          True  
3  VenusaurMegaVenusaur          True  
4            Charmander          True

In [194]:
# 이름에 알파벳이 아닌 다른 문자가 포함된 포켓몬 확인하기 - 포켓몬 9마리
print(pokemon[pokemon["name_isalpha"] == False].shape)
pokemon[pokemon["name_isalpha"] == False]

(9, 17)


#           Name   Type 1  Type 2  Total   HP  Attack  Defense  \
34    29      Nidoran X   Poison     NaN    275   55      47       52   
37    32      Nidoran Y   Poison     NaN    273   46      57       40   
90    83      Farfetchd   Normal  Flying    352   52      65       55   
131  122        Mr Mime  Psychic   Fairy    460   40      45       65   
252  233    Porygon Two   Normal     NaN    515   85      80       90   
270  250          Ho Oh     Fire  Flying    680  106     130       90   
487  439        Mime Jr  Psychic   Fairy    310   20      25       45   
525  474      Porygon Z   Normal     NaN    535   85      80       70   
794  718  Zygarde Forme   Dragon  Ground    600  108     100      121   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  name_count  long_name  \
34        40       40     41           1      False           8      False   
37        40       40     50           1      False           8      False   
90        58       62     60           1      False          10       True   
131      100      120     90           1      False           8      False   
252      105       95     60           2      False           8      False   
270      110      154     90           2       True           5      False   
487       70       90     60           4      False           8      False   
525      135       75     90           4      False           9      False   
794       81       95     95           6       True          16       True   

     Name_nospace  name_isalpha  
34       NidoranX         False  
37       NidoranY         False  
90      Farfetchd         False  
131        MrMime         False  
252    PorygonTwo         False  
270          HoOh         False  
487        MimeJr         False  
525      PorygonZ         False  
794  ZygardeForme         False

In [ ]:
# 해당 9마리 포켓몬의 이름 변경하기 - 알파벳이 아닌 것을 포함하는 이름 변경
pokemon = pokemon.replace(to_replace="Nidoran♀", value="Nidoran X")
pokemon = pokemon.replace(to_replace="Nidoran♂", value="Nidoran Y")
pokemon = pokemon.replace(to_replace="Farfetch'd", value="Farfetchd")
pokemon = pokemon.replace(to_replace="Mr. Mime", value="Mr Mime")
pokemon = pokemon.replace(to_replace="Porygon2", value="Porygon Two")
pokemon = pokemon.replace(to_replace="Ho-oh", value="Ho Oh")
pokemon = pokemon.replace(to_replace="Mime Jr.", value="Mime Jr")
pokemon = pokemon.replace(to_replace="Porygon-Z", value="Porygon Z")
pokemon = pokemon.replace(to_replace="Zygarde50% Forme", value="Zygarde Forme")

pokemon.loc[[34, 37, 90, 131, 252, 270, 487, 525, 794]]

In [ ]:
# 변경된 컬럼 Name으로 컬럼 Name_nospace를 업데이트하고, 알파벳이 아닌 것을 포함하는 이름 학인하기
pokemon["Name_nospace"] = pokemon["Name"].apply(lambda i: i.replace(" ", ""))

pokemon[pokemon["name_isalpha"] == False]

In [207]:
import re

# 토큰화 함수 만들기
def tokenize(name):
    name_split = name.split(" ")
    tokens = []
    for part_name in name_split:
        a = re.findall('[A-Z][a-z]*', part_name)
        tokens.extend(a)

    return np.array(tokens)

In [208]:
# 전설의 포켓몬의 이름에 사용된 토큰 확인하기 - 65개의 토큰
all_tokens = list(legendary["Name"].apply(tokenize).values)

token_set = []
for token in all_tokens:
    token_set.extend(token)

print(len(set(token_set)))
print(token_set)

65
['Articuno', 'Zapdos', 'Moltres', 'Mewtwo', 'Mewtwo', 'Mega', 'Mewtwo', 'X', 'Mewtwo', 'Mega', 'Mewtwo', 'Y', 'Raikou', 'Entei', 'Suicune', 'Lugia', 'Ho', 'Regirock', 'Regice', 'Registeel', 'Latias', 'Latias', 'Mega', 'Latias', 'Latios', 'Latios', 'Mega', 'Latios', 'Kyogre', 'Kyogre', 'Primal', 'Kyogre', 'Groudon', 'Groudon', 'Primal', 'Groudon', 'Rayquaza', 'Rayquaza', 'Mega', 'Rayquaza', 'Jirachi', 'Deoxys', 'Normal', 'Forme', 'Deoxys', 'Attack', 'Forme', 'Deoxys', 'Defense', 'Forme', 'Deoxys', 'Speed', 'Forme', 'Uxie', 'Mesprit', 'Azelf', 'Dialga', 'Palkia', 'Heatran', 'Regigigas', 'Giratina', 'Altered', 'Forme', 'Giratina', 'Origin', 'Forme', 'Darkrai', 'Shaymin', 'Land', 'Forme', 'Shaymin', 'Sky', 'Forme', 'Arceus', 'Victini', 'Cobalion', 'Terrakion', 'Virizion', 'Tornadus', 'Incarnate', 'Forme', 'Tornadus', 'Therian', 'Forme', 'Thundurus', 'Incarnate', 'Forme', 'Thundurus', 'Therian', 'Forme', 'Reshiram', 'Zekrom', 'Landorus', 'Incarnate', 'Forme', 'Landorus', 'Therian', 'Form

In [209]:
from collections import Counter

# 가장 많이 쓰인 토큰 확인하기
most_common = Counter(token_set).most_common(10)
most_common

[('Forme', 15),
 ('Mega', 6),
 ('Mewtwo', 5),
 ('Kyurem', 5),
 ('Deoxys', 4),
 ('Hoopa', 4),
 ('Latias', 3),
 ('Latios', 3),
 ('Kyogre', 3),
 ('Groudon', 3)]

In [211]:
# 가장 많이 쓰인 토큰 10개 컬럼 추가하기 - 전설의 포켓몬 이름에 등장하는 토큰이 포켓몬 이름에 있는지 확인
for token, _ in most_common:
    # pokemon[token] = ... 형식으로 사용하면 뒤에서 warning이 발생합니다
    pokemon[f"{token}"] = pokemon["Name"].str.contains(token)

pokemon.head(10)

#                       Name Type 1  Type 2  Total  HP  Attack  Defense  \
0  1                  Bulbasaur  Grass  Poison    318  45      49       49   
1  2                    Ivysaur  Grass  Poison    405  60      62       63   
2  3                   Venusaur  Grass  Poison    525  80      82       83   
3  3      VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
4  4                 Charmander   Fire     NaN    309  39      52       43   
5  5                 Charmeleon   Fire     NaN    405  58      64       58   
6  6                  Charizard   Fire  Flying    534  78      84       78   
7  6  CharizardMega Charizard X   Fire  Dragon    634  78     130      111   
8  6  CharizardMega Charizard Y   Fire  Flying    634  78     104       78   
9  7                   Squirtle  Water     NaN    314  44      48       65   

   Sp. Atk  Sp. Def  ...  Forme   Mega  Mewtwo  Kyurem  Deoxys  Hoopa  Latias  \
0       65       65  ...  False  False   False   False   False  False   False   
1       80       80  ...  False  False   False   False   False  False   False   
2      100      100  ...  False  False   False   False   False  False   False   
3      122      120  ...  False   True   False   False   False  False   False   
4       60       50  ...  False  False   False   False   False  False   False   
5       80       65  ...  False  False   False   False   False  False   False   
6      109       85  ...  False  False   False   False   False  False   False   
7      130       85  ...  False   True   False   False   False  False   False   
8      159      115  ...  False   True   False   False   False  False   False   
9       50       64  ...  False  False   False   False   False  False   False   

   Latios  Kyogre  Groudon  
0   False   False    False  
1   False   False    False  
2   False   False    False  
3   False   False    False  
4   False   False    False  
5   False   False    False  
6   False   False    False  
7   False   False    False  
8   False   False    False  
9   False   False    False  

[10 rows x 27 columns]

In [212]:
# Type1 및 Type2 종류 수 확인하기
len(list(set(pokemon["Type 1"]))), len(list(set(pokemon["Type 2"])))

(18, 19)

In [213]:
# Type1과 Type2의 차이 확인 - NaN 데이터 외 나머지 18가지 속성이 동일함
set(pokemon["Type 2"]) - set(pokemon["Type 1"])

{nan}

In [214]:
# 포켓몬들의 모든 Type 속성을 변수 types에 담기
types = list(set(pokemon["Type 1"]))
print(len(types))
print(types)

18
['Ground', 'Fire', 'Ghost', 'Steel', 'Flying', 'Fighting', 'Dragon', 'Bug', 'Electric', 'Rock', 'Fairy', 'Poison', 'Normal', 'Psychic', 'Grass', 'Ice', 'Dark', 'Water']


In [216]:
# 원-핫 인코딩 작업하기 - types에 담긴 18개의 속성 (컬럼)
for t in types:
    pokemon[t] = (pokemon["Type 1"] == t) | (pokemon["Type 2"] == t)

pokemon[[["Type 1", "Type 2"] + types][0]].head()

Type 1  Type 2  Ground   Fire  Ghost  Steel  Flying  Fighting  Dragon  \
0  Grass  Poison   False  False  False  False   False     False   False   
1  Grass  Poison   False  False  False  False   False     False   False   
2  Grass  Poison   False  False  False  False   False     False   False   
3  Grass  Poison   False  False  False  False   False     False   False   
4   Fire     NaN   False   True  False  False   False     False   False   

     Bug  Electric   Rock  Fairy  Poison  Normal  Psychic  Grass    Ice  \
0  False     False  False  False    True   False    False   True  False   
1  False     False  False  False    True   False    False   True  False   
2  False     False  False  False    True   False    False   True  False   
3  False     False  False  False    True   False    False   True  False   
4  False     False  False  False   False   False    False  False  False   

    Dark  Water  
0  False  False  
1  False  False  
2  False  False  
3  False  False  
4  False  False

# **(3) 베이스라인 모델 학습 성공 및 분석**

In [218]:
# 전처리된 모든 컬럼 확인하기 - 13개의 특성 (컬럼) -> 45개의 특성 (컬럼)
print(len(pokemon.columns))
print(pokemon.columns)

45
Index(['#', 'Name', 'Type 1', 'Type 2', 'Total', 'HP', 'Attack', 'Defense',
       'Sp. Atk', 'Sp. Def', 'Speed', 'Generation', 'Legendary', 'name_count',
       'long_name', 'Name_nospace', 'name_isalpha', 'Forme', 'Mega', 'Mewtwo',
       'Kyurem', 'Deoxys', 'Hoopa', 'Latias', 'Latios', 'Kyogre', 'Groudon',
       'Ground', 'Fire', 'Ghost', 'Steel', 'Flying', 'Fighting', 'Dragon',
       'Bug', 'Electric', 'Rock', 'Fairy', 'Poison', 'Normal', 'Psychic',
       'Grass', 'Ice', 'Dark', 'Water'],
      dtype='object')


In [219]:
# 모델에 사용하지 않을 컬럼 제외하고 features에 담기 - ID number, Name, name_nospace, name_isalpha, Type1, Type2, Legendary 제외
features = ['Total', 'HP', 'Attack', 'Defense','Sp. Atk', 'Sp. Def', 'Speed', 'Generation',
            'name_count','long_name', 'Forme', 'Mega', 'Mewtwo','Deoxys', 'Kyurem', 'Latias', 'Latios',
            'Kyogre', 'Groudon', 'Hoopa','Poison', 'Ground', 'Flying', 'Normal', 'Water', 'Fire',
            'Electric','Rock', 'Dark', 'Fairy', 'Steel', 'Ghost', 'Psychic', 'Ice', 'Bug', 'Grass', 'Dragon', 'Fighting']

len(features)

38

In [222]:
# 모델이 맞혀야 할 정답 데이터 준비하기 - 포켓몬 800마리의 전설의 포켓몬 여부
target = 'Legendary'
target

'Legendary'

In [223]:
# features에 해당하는 데이터를 변수 X에 담고 확인하기 - 포켓몬 800마리 및 각 포켓몬의 38가지 특성
X = pokemon[features]
print(X.shape)
X.head()

(800, 38)


Total  HP  Attack  Defense  Sp. Atk  Sp. Def  Speed  Generation  \
0    318  45      49       49       65       65     45           1   
1    405  60      62       63       80       80     60           1   
2    525  80      82       83      100      100     80           1   
3    625  80     100      123      122      120     80           1   
4    309  39      52       43       60       50     65           1   

   name_count  long_name  ...   Dark  Fairy  Steel  Ghost  Psychic    Ice  \
0           9      False  ...  False  False  False  False    False  False   
1           7      False  ...  False  False  False  False    False  False   
2           8      False  ...  False  False  False  False    False  False   
3          21       True  ...  False  False  False  False    False  False   
4          10       True  ...  False  False  False  False    False  False   

     Bug  Grass  Dragon  Fighting  
0  False   True   False     False  
1  False   True   False     False  
2  False   True   False     False  
3  False   True   False     False  
4  False  False   False     False  

[5 rows x 38 columns]

In [225]:
# 정답인 target에 해당하는 데이터를 변수 y에 담고 확인하기 - 포켓몬 800마리의 전설의 포켓몬 여부
y = pokemon[target]
print(y.shape)
y.head()

(800,)


0    False
1    False
2    False
3    False
4    False
Name: Legendary, dtype: bool

In [226]:
# 훈련 데이터와 테스트 데이터 분리하기 - 훈련 데이터 640개 - 테스트 데이터 160개
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=15)

print(X_train.shape, y_train.shape)
print(X_test.shape, y_test.shape)

(640, 38) (640,)
(160, 38) (160,)


In [227]:
# # 의사결정 트리 모델을 변수 model에 저장하기
model = DecisionTreeClassifier(random_state=25)
model

DecisionTreeClassifier(random_state=25)

In [228]:
# 훈련 데이터로 모델 훈련 시키기
model.fit(X_train, y_train)

# 테스트 데이터를 통해 예측값 얻기
y_pred = model.predict(X_test)

In [229]:
# # 예측값 (y_pred) 및 정답 (y_test) 확인하기
confusion_matrix(y_test, y_pred)

array([[141,   6],
       [  1,  12]])

## **✅ 작업 수행 결과**

            Positve: Legendary=True (전설의 포켓몬)   
            Negative: Legendary=False (일반 포켓몬)   
   
                - TN (True Negative): 141   
                    일반 포켓몬을 일반 포켓몬이라고 알맞게 판단
           
                - FP (False Positive): 6   
                    일반 포켓몬을 전설의 포켓몬이라고 잘못 판단
            
                - FN (False Negative): 1   
                    전설의 포켓몬을 일반 포켓몬이라고 잘못 판단
               
                - TP (True Positive): 12    
                    전설의 포켓몬을 전설의 포켓몬이라고 알맞게 판단    



              
- *총 160개의 테스트 데이터 중 7개만 잘못 판단하여 약 95.6%의 높은 정확도 (accuracy)를 달성하였다.*   
- *그러나, 전설의 포켓몬과 일반 포켓몬이 불균형하게 분포되어 있는 해당 데이터에서 정확도만으로 모델의 성능을 평가하기에는 무리가 있다.*      

In [230]:
# 다른 척도로 모델의 성능 평가하기
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

       False       0.99      0.96      0.98       147
        True       0.67      0.92      0.77        13

    accuracy                           0.96       160
   macro avg       0.83      0.94      0.87       160
weighted avg       0.97      0.96      0.96       160



## **✅ 작업 수행 결과**

- *데이터 전처리 과정을 거치지 않고 데이터를 선별하여 8개의 특성으로 모델을 훈련시키고 테스트한 Recall 결과가 0.62에 그친 것과 대조적이다.*    
- *데이터 전처리 과정을 거쳐 데이터를 선별하여 38개의 특성으로 모델을 훈련시키고 테스트한 Recall 결과가 0.92로까지 향상되었다.*     
- *따라서, 모델의 성능 향상을 위해서 데이터 전처리 과정의 중요성을 간과할 수 없다.*

# **(4) 비교 모델 학습 성공 및 비교 분석**

## **⭐ Random Forest**

In [231]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

from sklearn.ensemble import RandomForestClassifier  # 랜덤포레스트라는 분류기를 사용하기 위해 import

random_forest = RandomForestClassifier(random_state=32)  # RandomForest분류기 객체 생성
print(random_forest._estimator_type)  # 이 모델의 타입 확인
print()

random_forest.fit(X_train, y_train) # 훈련

classifier



RandomForestClassifier(random_state=32)

In [232]:
y_pred = random_forest.predict(X_test) # 예측

print("Random Forest Classification Report:")
print(classification_report(y_test, y_pred)) # 결과 지표를 확인

Random Forest Classification Report:
              precision    recall  f1-score   support

       False       0.99      0.97      0.98       147
        True       0.73      0.85      0.79        13

    accuracy                           0.96       160
   macro avg       0.86      0.91      0.88       160
weighted avg       0.97      0.96      0.96       160



## **⭐ SVM (Support Vector Machine)**

In [233]:
from sklearn import svm  # Support Vector Machine을 사용하기 위해 import

svm_model = svm.SVC()  # 모델 객체 생성
print(svm_model._estimator_type)  # 이 모델의 타입 확인
print()

svm_model.fit(X_train, y_train)  # 훈련

classifier



SVC()

In [234]:
y_pred = svm_model.predict(X_test)  # 예측

print("SVM Classification Report:")
print(classification_report(y_test, y_pred)) # 결과 지표를 확인

SVM Classification Report:
              precision    recall  f1-score   support

       False       0.96      0.99      0.97       147
        True       0.78      0.54      0.64        13

    accuracy                           0.95       160
   macro avg       0.87      0.76      0.80       160
weighted avg       0.95      0.95      0.95       160



## **⭐ SGD Classifier**

In [235]:
from sklearn.linear_model import SGDClassifier  # 선형분류기인 SGDClassifier를 사용하기 위한 import

sgd_model = SGDClassifier()  # 모델 객체 생성
print(sgd_model._estimator_type)  # 이 모델의 타입 확인
print()

sgd_model.fit(X_train, y_train)  # 훈련

classifier



SGDClassifier()

In [236]:
y_pred = sgd_model.predict(X_test) # 예측

print("SGD Classifier Classification Report:")
print(classification_report(y_test, y_pred)) # 결과 지표를 확인

SGD Classifier Classification Report:
              precision    recall  f1-score   support

       False       0.93      0.94      0.94       147
        True       0.25      0.23      0.24        13

    accuracy                           0.88       160
   macro avg       0.59      0.58      0.59       160
weighted avg       0.88      0.88      0.88       160



## **⭐ Logistic Regression**

In [237]:
from sklearn.linear_model import LogisticRegression  # 선형분류기인 LogisticRegression를 사용하기 위한 import

logistic_model = LogisticRegression()  # 모델 객체 생성
print(logistic_model._estimator_type)  # 이 모델의 타입 확인
print()

logistic_model.fit(X_train, y_train)  # 훈련

classifier



/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:460: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [238]:
y_pred = logistic_model.predict(X_test)  # 예측

print("Logistic Regression Classification Report:")
print(classification_report(y_test, y_pred))  # 결과 지표 확인

Logistic Regression Classification Report:
              precision    recall  f1-score   support

       False       0.95      1.00      0.98       147
        True       1.00      0.46      0.63        13

    accuracy                           0.96       160
   macro avg       0.98      0.73      0.80       160
weighted avg       0.96      0.96      0.95       160



## **✅ 작업 수행 결과**

- *Random Forest에서 3가지 지표가 모두 높게 나타나면서, 불균형 데이터가 잘 처리되고 다양한 특징이 잘 학습될 것으로 보인다.*             



- Decision Tree Classifier     
    정밀도: 0.67       
    재현율: 0.92     
    F1-score: 0.77     

- Random Forest     
    정밀도: 0.73   
    재현율: 0.85   
    F1-score: 0.79   

- SVM (Support Vector Machine)     
    정밀도: 0.78   
    재현율: 0.54   
    F1-score: 0.64   

- SGD Classifier     
    정밀도: 0.25   
    재현율: 0.23   
    F1-score: 0.24   

- Logistic Regression     
    정밀도: 1.00   
    재현율: 0.46   
    F1-score: 0.63   



## **✅ 총평**

- *데이터가 불균형하므로 단순히 정확도 (accuracy)를 사용하는 것 보다는 다른 평가 지표를 사용하는 것이 중요하다.*        
- *전설의 포켓몬이 소수이기 때문에, 주로 정밀도, 재현율, F1-score 등의 성능 지표를 사용해야 한다.*

# **✅ 보충 설명**

- Decision Tree Classifier     
    - 간단한 규칙 기반의 분류 모델   
    - 해석이 쉬움   
    - 과적합 가능성 있음    
    - 소수 클래스에 민감    
    - 데이터에 노이즈가 있을 경우 성능 저하   

- Random Forest   
    - 여러 개의 Decision Tree를 조합   
    - Decision Tree보다 일반적으로 더 좋은 성능      
    - 과적합에 덜 민감      
    - 불균형 데이터에서도 강력한 성능     

- SVM (Stochastic Gradient Descent)     
    - 대규모 데이터셋에 적합
    - 간단하고 빠름       
    - 적절한 하이퍼파라미터 튜닝 필요       
    - 데이터가 불균형하면 성능 저하        

- Logistic Regression.    
    - 이진 분류 문제에서 강력한 성능      
    - 선형 모델       
    - 데이터가 선형적으로 분리 가능하면 성능 좋음       
    - 불균형 데이터에 대해서는 성능이 좋지 않음         
    - 클래스 가중치 조정 가능        


            정밀도: 모델이 전설의 포켓몬이라고 예측한 것 중에서 실제로 전설의 포켓몬인 비율     

            재현율: 실제 전설의 포켓몬 중 모델이 정확하게 예측한 비율     

            F1-score: 정밀도와 재현율의 조화 평균으로, 불균형 데이터에서 중요한 평가 지표          


          

